### DISCUSSION TITLE
API endpoint for repo social media preview?

### DISCUSSION THREAD

**Discussion by Andrew6rant** [community/community#32166] [Apps, API and Webhooks]
Hello, I was wondering if it was possible to change a repository's social media preview image (https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/customizing-your-repositorys-social-media-preview) without going through the GitHub website.

I didn't see anything in the GraphQL API nor the REST API, but I may have missed it.

Thanks,

Andrew Grant

**Comment by karol-brejna-i** [comment]
Same here. I went through GraphQL an REST APIs documentation. No luck. 
Is there a way to bring this to the GitHub dev team as a feature request?

**Comment by dpuga1** [comment]
I’m really new to all this I couldn’t tell u.
David Puga

**Comment by ShivamJoker** [comment]
@Andrew6rant I think this should be a feature request, may I know why you marked this as answered?

**Reply by Andrew6rant** [reply]
Averagebit emailed me asking if I could mark their response as answered. It looked like there is no API endpoint, so I figured, why not?

If there's a specific way for me to request this feature, I'm all ears

**Reply by ShivamJoker** [reply]
I was told to ask here for the feature request, it should only be marked answered once we have the API I believe.

**Reply by Andrew6rant** [reply]
Alright, I've un-marked the comment.

**Comment by SableRaf** [comment]
I built [a node.js tool](https://github.com/nestofbirbs/opengraph-image) that creates customizable OG images in the style of GitHub's own social media previews. It replicates the technique described on the GitHub blog in [A framework for building Open Graph images](https://github.blog/open-source/git/framework-building-open-graph-images/). I was hoping to automate updates via GitHub actions but it looks like this will not be possible 🥲 I hope this feature gets implemented in the API some day so I can revisit this project!

![image](https://github.com/user-attachments/assets/ebf332fa-8ee9-4338-92f7-836782bee4d9)

**Reply by metaory** [reply]
@SableRaf good work,

It can be much simpler..

Without [pptr](https://pptr.dev/)

Here is my naive **`JIT`** design,

#### TL;DR

```mermaid
---
config:
  layout: elk
  theme: base
  themeVariables:
    lineColor: '#EE5588'
    mainBkg: '#EE5588'
    textColor: '#FFF'
---
flowchart LR
    Req(Request) -- GET --> R(Route)
    R --> C{"R2"}
    C -- HIT --> F(Fetch)
    C -- MISS --> Col(Collect)
    F --> Res(Respond)
    Col --> Pop(Populate)
    Pop --> Ras(Rasterize)
    Ras --> S(Store)
    S --> Res
```

---

> You can replace Cloudflare with AWS or other providers

### Services
- CF Workerd dynamic route
- ~~CF KV with some TTL~~
- CF R2+lifecycle

### Cache Miss
1. Dynamic route hit
2. Cache Miss
3. Collect data (gql queries)
4. Populate SVG template
5. Rasterize SVG
6. Store PNG
7. Prepare response

---

_Looking at the docs, [SVG is not supported](https://indieweb.org/The-Open-Graph-protocol#Does_not_support_SVG_images) and base64 data urls are discouraged,
So cant go with KV and data url b64_

#### Some ref points:

##### CF
- [github.com/kvnang/workers-og](https://github.com/kvnang/workers-og)
- [developers.cloudflare.com/workers/tutorials/generate-dynamic-og-images-using-workers](https://developers.cloudflare.com/workers/tutorials/generate-dynamic-og-images-using-workers)
- [og-image-generator-cloudflare-worker](https://github.com/mohdlatif/og-image-generator-cloudflare-worker/tree/main)

which all use `@vercel/og` and `@vercel/satori` 
**not for very good reason, it can be simpler in vanilla without supporting `jsx`**

##### Vercel
- [npmjs.com/package/@vercel/og](https://npmjs.com/package/@vercel/og)
  - [github.com/vercel/next.js/packages/next/src/server/og/image-response.ts](https://github.com/vercel/next.js/blob/canary/packages/next/src/server/og/image-response.ts)
- [@vercel/satori](https://github.com/vercel/satori) _convert HTML and CSS to SVG_

**Reply by SableRaf** [reply]
Thanks for the suggestion—it’s an interesting approach, especially for scenarios where real-time or dynamic OG image generation is needed. 

My project is focused on generating a static OG image for a repo using a GitHub Action. Since repository metadata doesn’t change that often, it would be good enough and more efficient to run it on a cron job, say once a day.

The main limitation I’m working around is that GitHub doesn’t have an API for updating repo OG images automatically (the topic of the discussion above). The dynamic infrastructure you suggested is not really applicable here since GitHub repos do not allow JIT rendering of OG images (unless I missed something huge).

That said, I can see how your design would work well in other contexts like a social media platform, or blogs with frequently changing content. Thanks again for sharing your thoughts!

**Reply by metaory** [reply]
Oops my bad, my ADHD brain got tunnel visioned to what it found interesting and completely ignored all context..

I'm sorry..

 @SableRaf  

Reading the original post again I now see what the goal was..

This is sth I was trying to achieve recently as well, and gave up midway,

Yes there is no public API to update repository image preview.

So I started reverse engineering what GitHub client is doing, piecing together their client code from sourcemaps and reconstructing the request they make.

This is the endpoint;  writing to S3
`POST:https://github.com/upload/policies/repository-images`

Reading these we can reconstruct the request,
- `modules/github/settings/upload-repository-image.ts`
- `modules/github/upload/batch-upload.ts`
- `modules/github/upload/attachment-upload.ts`

You can let devtool unwrap them to debug,

Here some tool and ref to explore sourcemaps with relevant github code;
- [settings-source-map](https://github.githubassets.com/assets/settings-fe0ee97d5327.js.map) 
- [settings-source-map-visuals](https://metaory.github.io/source-map-visualization/?url=https://github.githubassets.com/assets/settings-fe0ee97d5327.js.map)


Lots of funny stuff they doing,, I stopped here 

---

Btw I don't understand why GitHub not moving all these metadata to git files,

Here is how they could've done and be a good Linux citizen

1. GitHub repository description 
`.git/description`

2. GitHub repository topics
`.git/info/topics`

3. GitHub repository social banner
`.git/info/social.png`

...


I'm daydreaming again 🙃

Have a great day 🦉

**Reply by metaory** [reply]
stubborn disregard for git internals like `.git/description` from gh with nih syndrome .. 🫥

**Reply by c260fg** [reply]
may not be the best place to ask but is it configurable to use it with any input picture and not an auto-generated one?

**Reply by SableRaf** [reply]
Yes @c260fg, everything is configurable. If you don't mind updating the image manually, you should be able to use the tool. 

Check out the README for details. 

https://github.com/nestofbirbs/opengraph-image

**Comment by Builder106** [comment]
Hey, refreshing this in 2026 with some agentic-workflow context — see #197021. I've added three new use cases (Claude Code scaffolding, the MCP-server tooling layer, security-tooling release pipelines). Hoping this feature gets added 🙏

**Comment by hyperpolymath** [comment]
Strong +1. The gap is especially visible for teams managing **branding-as-code**: `description`, `homepage`, and `topics` are all API-settable, so repositories can be provisioned and kept in sync automatically — but the social preview still requires a manual click in Settings, which is the one thing that breaks otherwise fully-automated repo bootstrap and org-wide branding rollouts.

A `PUT /repos/{owner}/{repo}/social-preview` accepting an image upload (PNG/JPG/GIF, ≤1 MB), or an `og_image` field on `PATCH /repos/{owner}/{repo}`, would close it. Committing an image into the repo does **not** populate the Settings slot, so there is currently no automatable path at all.

**Comment by ketpatil77** [comment]
An API for repository social preview images would unlock a lot of automation use cases that are currently awkward or impossible: release branding, docs-site refreshes, multilingual marketing assets, and consistent preview generation across many repos. Since the preview image is already repository configuration, exposing it through REST or GraphQL feels aligned with how other repo-level metadata is managed. Even a minimal create/update/delete endpoint would be useful, especially if it worked cleanly with GitHub Actions-driven branding workflows.
